### PR TITLE
Feature: Add legacy support and Union schema compatibility

### PR DESCRIPTION
**Pull request by fivetran-catfritz**
## PR Overview
**This PR will address the following Issue/Feature:**
- [#23 ](https://github.com/fivetran/dbt_lever_source/issues/23)
- [#25 ](https://github.com/fivetran/dbt_lever_source/issues/25)

**This PR will result in the following new package version:**

- v0.7.0 due to schema changes

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

> ## Features
> - For Fivetran Lever connectors created on or after July 27, 2024, the `USER` and `INTERVIEWER_USER` source tables have been renamed to `USERS` and `INTERVIEW_USER`, respectively. This package now prioritizes the `USERS` and `INTERVIEW_USER` tables if available, falling back to `USER` and `INTERVIEWER_USER` if not.
>   - To continue using the `USER` and/or `INTERVIEWER_USER` tables, set the variables `lever__using_users` and/or `lever__using_interview_user` to false in your `dbt_project.yml`.
>   - For more information, refer to the [July 2024 connector release notes](https://fivetran.com/docs/connectors/applications/lever/changelog#july2024) and the related [README section](https://github.com/fivetran/dbt_lever/blob/main/README.md##leveraging-legacy-connector-table-names).
> 
> - Introduced the ability to union source data from multiple Lever connectors. For more details, see the related [README section](https://github.com/fivetran/dbt_lever/blob/main/README.md#union-multiple-connectors).
> 
> ## Bug fixes
> - Fixed an issue where the dbt package would error due to a missing `CONTACT_LINK` source table for users without source data, even though it was enabled in the Fivetran Connector. A null-filled table will now be generated in such cases.
> 
> ## Under the hood
> - In the source package, updated temporary models to union source data using the `fivetran_utils.union_data` macro.
> - Added the `source_relation` column in each staging model to identify the origin of each field, utilizing the `fivetran_utils.source_relation` macro.
> - Updated tests to include the new `source_relation` column.
> - Included the `source_relation` column in all joins and window function partition clauses in the transform package.
> - Added consistency tests for end models.

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [ ] ~dbt run (if incremental models are present) && dbt test~

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked, tagged, and properly assigned
- [x] All necessary documentation and version upgrades have been applied
- [x] docs were regenerated (unless this PR does not include any code or yml updates)
- [x] BuildKite integration tests are passing
- [x] Detailed validation steps have been provided below

### Detailed Validation
Please share any and all of your validation steps:

- Similar updates as found in [THIS](https://github.com/fivetran/dbt_recharge_source/pull/12) Recharge PR.

- Validation consistency tests pass when using below exclusion var. Source relation should be excluded this run since it is a new field.
```yml
var:
  consistency_test_exclude_metrics: ['source_relation']
```
Screenshot 2024-10-21 at 5 23 34 PM

### If you had to summarize this PR in an emoji, which would it be?
 
:dancer:

### CODE REVIEW

File: README.md
Comment (fivetran-reneeli) [inline README.md:1]: dont forget to up the version in line #53 and 131!

**Review by fivetran-reneeli — COMMENTED** [review]
Thanks @fivetran-catfritz ! Just had some comments

**Review by fivetran-catfritz — COMMENTED** [review]
@fivetran-reneeli Thanks for reviewing! I have made the suggested updates.

**Review by fivetran-reneeli — APPROVED** [review]
Thanks @fivetran-catfritz ! Just reminder to uncollapse the dropdowns in the README. Other than that, looks good to go!

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,23 @@
+# dbt_lever v0.7.0
```
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:1]: Request to add some of the comments from the source PR to this one as well so users are aware of those upstream changes as well.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,23 @@
...
+## Bug fixes
+- Fixed an issue where the dbt package would error due to a missing `CONTACT_LINK` source table for users without source data, even though it was enabled in the Fivetran Connector. A null-filled table will now be generated in such cases.
+
+## Under the hood
+- In the source package, updated temporary models to union source data using the `fivetran_utils.union_data` macro.
+- Added the `source_relation` column in each staging model to identify the origin of each field, utilizing the `fivetran_utils.source_relation` macro.
+- Updated tests to include the new `source_relation` column.
+- Included the `source_relation` column in all joins and window function partition clauses in the transform package.
+- Added consistency tests for end models.
+- Remove unnecessary tests for intermediate models to reduce resource usage.
```
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:19]: I really like this addition! No need for unnecessary tests. Thanks for thinking proactive here and reducing this debt!

File: models/lever.yml [hunk 3]
```diff
@@ -87,12 +89,16 @@ models:
...
+            - source_relation 
+            - opportunity_id
+    columns: 
+      - name: source_relation
+        description: "{{ doc('source_relation') }}"
       - name: opportunity_id
         description: Unique ID of the opportunity.
-        tests:
-          - unique
-          - not_null
```
Comment (fivetran-joemarkiewicz) [inline models/lever.yml:95]: Was this not_null intentionally removed? If so, we should add this to the CHANGELOG. Otherwise, we should add it back in as a test on this field.

Comment (fivetran-joemarkiewicz) [reply models/lever.yml:95]: Same note for other removals

File: packages.yml [hunk 4]
```diff
@@ -1,3 +1,6 @@
 packages:
-  - package: fivetran/lever_source
-    version: [">=0.6.0", "<0.7.0"]
\ No newline at end of file
+  # - package: fivetran/lever_source
+  #   version: [">=0.7.0", "<0.8.0"]
+- git: https://github.com/fivetran/dbt_lever_source.git
+  revision: bug/add-legacy-support
+  warn-unpinned: false
```
Comment (fivetran-joemarkiewicz) [inline packages.yml:6]: Reminder to swap before release

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-catfritz thanks for working though these updates! I have a few change requests before this is ready for approval. Let me know if you have any questions. Thanks!

**Review by fivetran-catfritz — COMMENTED** [review]
@fivetran-joemarkiewicz Thank you and made the updates!

**Review by fivetran-catfritz — COMMENTED** [review]
@fivetran-joemarkiewicz I added back the null tests and opened the issues for the affected packages. Ready for re-review!

File: models/lever.yml [hunk 5]
```diff
@@ -87,11 +101,17 @@ models:
 
   - name: lever__opportunity_enhanced
     description: Table of unique opportunities.
-    columns:
+    tests:
+      - dbt_utils.unique_combination_of_columns:
+          combination_of_columns:
+            - source_relation 
+            - opportunity_id
```
Comment (fivetran-joemarkiewicz) [inline models/lever.yml:108]: This is a bit out of scope, but I believe we should add `application_id` to this uniqueness test. When running with some internal data I receive a test failure on this model because of individuals who applied to the same opportunity multiple times.

It looks like this is a known phenomena as well since [we explicitly call this behavior out](https://github.com/fivetran/dbt_lever_source/blob/6d605f9a0462c0e45307fa18dbcbb9c40b106ceb/models/stg_lever.yml#L5-L8) in the staging model for opportunity. It then must come to fruition once we do [this join](https://github.com/fivetran/dbt_lever/blob/850b6928bf054d186634ec6be3cca1279bbe2996/models/intermediate/int_lever__opportunity_application.sql#L27-L30).

Would you be able to look into this a bit more and either add the `application_id` to the uniqueness test if it isn't too out of scope. Let me know if this ends up breaching too far out of scope, but if we can it would be great to address this alongside this next release. Thanks!

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,27 @@
...
+- Add missing field definitions to dbt docs.
+
+## Under the hood
+- In the source package, updated temporary models to union source data using the `fivetran_utils.union_data` macro.
+- Added the `source_relation` column in each staging model to identify the origin of each field, utilizing the `fivetran_utils.source_relation` macro.
+- Updated tests to include the new `source_relation` column.
+- Included the `source_relation` column in all joins and window function partition clauses in the transform package.
+- Added consistency tests for end models.
+- Turned off freshness tests for the `USER`/`USERS` and `INTERVIEW_USER`/`INTERVIEWER_USER` sources to avoid possible conflicts.
+- Remove unnecessary tests for intermediate models to reduce resource usage.
```
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:23]: Similar to the comment from the source package, can you add a blanket entry to say while you removed the tests from the int models, you did add tests to the end models and made sure they all had the appropriate unique and not null test coverage.

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-catfritz thanks for working through this PR! We're almost there!! I have a few final comments and one question I would like you to look a bit further into regarding a failing test. I do want to mention this test is failing for the current version of Lever as well, so this is not a new bug but one we or others haven't seemed to report yet. I would like to look into this if possible.

File: models/lever.yml (hunk 5 above)
Comment (fivetran-catfritz) [reply models/lever.yml:108]: I opened [this issue ](https://github.com/fivetran/dbt_lever/issues/22) since I had to investigate if the model or the test had to be updated. I can look at it next if we're ok delaying release.

File: models/lever.yml [hunk 7]
```diff
@@ -42,6 +52,8 @@ models:
         description: Name of timezone in which interview was scheduled to occur.
       - name: feedback_form_id
         description: Unique id of the feedback form. 
+        tests:
+          - not_null
```
[the remark this answers is not in the document]
Comment (fivetran-catfritz) [reply models/lever.yml]: Ahh thank you. Updated and regenned docs.

File: models/lever.yml (hunk 5 above)
Comment (fivetran-joemarkiewicz) [reply models/lever.yml:108]: Ah yes thank you, I remember you opening this issue and raising it to me previously. Apologies for missing it. Since this change (if it's just a test update) would likely be a patch update it should be fine to move this investigation to next week and get this change out to address the larger issue we're trying to address with this update.

Given the above, can we recategorize this issue as a Bug instead of a FR so we can make sure we prioritize it over other issues next sprint.

File: CHANGELOG.md (hunk 6 above)
Comment (fivetran-catfritz) [reply CHANGELOG.md]: Updated!

**Review by fivetran-catfritz — COMMENTED** [review]
@fivetran-joemarkiewicz Ready for re-review!

**Review by fivetran-joemarkiewicz — APPROVED** [review]
LGTM!

File: packages.yml (hunk 4 above)
Comment (fivetran-catfritz) [inline packages.yml:6]:
```suggestion
  - package: fivetran/lever_source
    version: [">=0.7.0", "<0.8.0"]
```
